### DISCUSSION TITLE
getSession from subdomain not working

### DISCUSSION THREAD

**Discussion by datteroandrea** [nextauthjs/next-auth#5500] [Help]
Hello, I am building a platform that offers different applications, the main platform is running on http://localhost and the applications will run each on a specific subdomain, at the moment I have an application running on http://sub.localhost.

I am using Nginx and Docker to host both the platform and the application, my goal would be to authenticate on http://localhost and use the session of the platform in the applications (subdomains), I have already taken a look at every single source/similar problem but could not find a solution, some of the sources I have read are the following:

- https://github.com/nextauthjs/next-auth/discussions/1299
- https://github.com/nextauthjs/next-auth/issues/405
- https://github.com/nextauthjs/next-auth/issues/2718
- https://github.com/vercel/examples/tree/main/solutions/subdomain-auth

At the moment this is my .env.local on the main platform:

```
GOOGLE_CLIENT_ID=...
GOOGLE_CLIENT_SECRET=...
GOOGLE_AUTH_URL=...

NEXTAUTH_URL=http://localhost/
NEXTAUTH_URL_INTERNAL=http://mygames:3000/
NEXTAUTH_SECRET=...
DATABASE_URL=...
NEXT_PUBLIC_API_KEY=...
NEXT_SECRET_API_KEY=...
```

The following is the .env.local of the application (subdomain):

```
NEXTAUTH_URL=http://sub.localhost/
NEXTAUTH_URL_INTERNAL=http://mygames:3000/
NEXTAUTH_SECRET=...
DATABASE_URL=...
NEXT_PUBLIC_API_KEY=...
NEXT_SECRET_API_KEY=...
```

The following is my [...nextauth].js for the main platform:

```JS
import NextAuth from 'next-auth';
import GoogleProvider from 'next-auth/providers/google';
import { PrismaAdapter } from '@next-auth/prisma-adapter';
import prisma from '../../../lib/prisma';
import Stripe from 'stripe';

const getDomainWithoutSubdomain = url => {
  const urlParts = new URL(url).hostname.split('.');

  return urlParts
      .slice(0)
      .slice(-(urlParts.length === 4 ? 3 : 2))
      .join('.');
};

const hostName = getDomainWithoutSubdomain(process.env.NEXTAUTH_URL);

console.log("HOSTNAME", hostName);

const options = {
  secret: process.env.NEXTAUTH_SECRET,
  adapter: PrismaAdapter(prisma),
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
  ],
  pages: {
    signIn: '/signin'
  },
  callbacks: {
    async signIn({ user, account, profile, email, credentials }) {
      return true;
    },
    async redirect({ url, baseUrl }) {
      return baseUrl;
    },
    async session({ session, user, token }) {
      return { ...session, ...user };
    },
    async jwt({ token, user, account, profile, isNewUser }) {
      return token;
    }
  },
  cookies: {
    sessionToken: {
      name: process.env.NODE_ENV === 'production' ? `__Secure-next-auth.session-token` : 'next-auth.session-token',
      options: {
        httpOnly: true,
        sameSite: 'lax',
        path: '/',
        secure: process.env.NODE_ENV === 'production' ? true: false,
        domain: '.' + hostName
      }
    }
  }
}

export default (req, res) => NextAuth(req, res, options)
```

When I use getSession in the subdomain application I receive a null object, what can I do to solve this? Feel free to ask for anything for more details!

**Comment by ThangHuuVu** [comment]
Can you try using the same `NEXTAUTH_URL` between the 2 apps? I'm guessing this is the reason 🤔
I'm turning this issue into a discussion because we already have a working example, the one you linked in the description: https://github.com/vercel/examples/tree/main/solutions/subdomain-auth so it's not likely a bug in NextAuth.js
I will reopen if proven otherwise! 👀

**Reply by datteroandrea** [reply]
Even though I moved on with a different approach (I am using subpaths atm), I remember that when I tried to use the same NEXTAUTH_URL it was still not working.

**Reply by jackrdye** [reply]
Can confirm it does not work! I think it has something to do with getSession being client side however the NEXTAUTH_URL is not accessible client side since front end frame works often require some prepend for client side environment varaibles. e.g. Next - NEXT_PUBLIC, react app - REACT_APP. Therefore when you read NEXTAUTH_URL it just returns null and default sends the api request to the current domain not the intented next auth domain! Please fix - provide another method for confirguring NEXTAUTH_URL like an options object etc !!!!!
